core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -375,1083 +380,674 @@ public boolean isPreloaded() {
-                     semaphore -> handler.blockingPublisher(segmentedStore.publishKeys(segments, filter)),
-                     Semaphore::release);
-      }
-
-      return publishKeys(PersistenceUtil.combinePredicate(segments, keyPartitioner, filter), predicate);
+   public <K, V> CompletionStage<MarshallableEntry<K, V>> loadFromAllStores(Object key, boolean localInvocation,
+         boolean includeStores) {
+      return loadFromAllStores(key, keyPartitioner.getSegment(key), localInvocation, includeStores);
    }
 
-   private <K, V> MarshallableEntry<K, V> loadFromAllStoresSync(Object key, boolean localInvocation, boolean includeStores, int traceId) {
-      acquireReadLock();
-      try {
-         checkStoreAvailability();
-         if (trace) {
-            log.tracef("Loading entry for key %s from stores with includeStores %s for id: %d",
-                  key, includeStores, traceId);
-         }
-         MarshallableEntry load = null;
-         for (CacheLoader l : loaders) {
-            if (allowLoad(l, localInvocation, includeStores)) {
-               load = l.loadEntry(key);
-               if (load != null)
-                  break;
-            }
-         }
-         if (trace) {
-            log.tracef("Entry was loaded: %s for key %s from stores for id: %d", load, key, traceId);
-         }
-         return load;
-      } finally {
-         releaseReadLock();
+   @Override
+   public <K, V> CompletionStage<MarshallableEntry<K, V>> loadFromAllStores(Object key, int segment,
+         boolean localInvocation, boolean includeStores) {
+      return Maybe.using(
+            this::acquireReadLock,
+            ignore -> {
+               checkStoreAvailability();
+               if (trace) {
+                  log.tracef("Loading entry for key %s with segment %d", key, segment);
+               }
+               return Flowable.fromIterable(stores.entrySet())
+                     .filter(entry ->
+                           !entry.getValue().characteristics.contains(Characteristic.WRITE_ONLY)
+                                 && allowLoad(entry.getValue(), localInvocation, includeStores))
+                     // Only do 1 request at a time
+                     .concatMapMaybe(entry -> Maybe.fromCompletionStage(
+                           PersistenceManagerImpl.<K, V>storeForEntry(entry).load(segment, key)), 1)
+                     .firstElement();
+            },
+            this::releaseReadLock
+      ).toCompletionStage(null);
+   }
+
+   private boolean allowLoad(StoreStatus storeStatus, boolean localInvocation, boolean includeStores) {
+      return (localInvocation || !isLocalOnlyLoader(storeStatus.store)) &&
+            (includeStores || storeStatus.characteristics.contains(Characteristic.READ_ONLY) || storeStatus.config.ignoreModifications());
+   }
+
+   private boolean isLocalOnlyLoader(NonBlockingStore store) {
+      if (store instanceof LocalOnlyCacheLoader) return true;
```

Comment:
The only implementation of `LocalOnlyCacheLoader` is `ClusterLoader` ... when is that ever used?

Reply:
I forget the use case of ClusterLoader, I have a feeling it was something with invalidation caches. @danberindei might remember. I would like to remove this if possible, but I don't think we can yet. I think another JIRA or discussion would be required first.

Reply:
The decision is to [deprecate](https://issues.redhat.com/browse/ISPN-11863) and then remove in [14](https://issues.redhat.com/browse/ISPN-11864).